src/core/programs.rkt

```diff
@@ -170,6 +172,51 @@
   (-> location? expr? expr? expr?)
   (location-do loc prog (const prog*)))
 
+(define/contract (batch-location-set loc0 mutable-batch cache full-batchref sub-batchref)
+  (-> location? mutable-batch? hash? batchref? batchref? batchref?)
+  (match-define (batchref sub-batch sub-idx) sub-batchref)
+  (match-define (batchref full-batch full-idx) full-batchref)
+  (unless (equal? sub-batch full-batch)
+    (error 'batch-location-set "Function assumes that batches are equal"))
+
+  (define (get-node idx)
+    (hash-ref cache idx (λ () (vector-ref (batch-nodes full-batch) idx))))
```

Comment:
I see what you did. My idea was to put the cache here inside `mutable-batch` and just refer to it every time you do `batch-get` and `batch-push` or whatever they're called.